File: common/src/main/java/com/ssblur/scriptor/helpers/generators/CommunityModeGenerator.java
package com.ssblur.scriptor.helpers.generators;

import com.google.gson.JsonObject;
import org.jetbrains.annotations.Nullable;

import java.util.Random;

public class CommunityModeGenerator extends TokenGenerator {
  protected static String[] consonantGroup = new String[]{
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
    "h",
    "j",
    "k",
    "l",
    "m",
    "n",
    "p",
    "r",
    "s",
    "t",
    "v",
    "w",
    "x",
    "z",
    "st",
    "th",
    "ld",
    "qu",
    "pr",
    "lk",
    "mm",
    "pp",
    "ck",
    "rd",
    "pt"
  };
  protected static String[] vowelGroup = new String[]{
    "a",
    "e",
    "i",
    "o",
    "u",
    "ou",
    "ao",
    "io",
    "ee"
  };

  public CommunityModeGenerator(JsonObject object) {}

  @Override
  public String generateToken(String key, @Nullable JsonObject parameters) {
    Random random = new Random(key.hashCode() + 0x055b10b0);
    StringBuilder builder = new StringBuilder();
    boolean consonantFirst = random.nextBoolean();
    for(int i = 0; i < 3 + random.nextInt(4); i++) {
      if (consonantFirst ^ (i % 2 == 1))
        builder.append(vowelGroup[random.nextInt(vowelGroup.length)]);
      else
        builder.append(consonantGroup[random.nextInt(consonantGroup.length)]);
    }

    return builder.toString();
  }

  @Override
  public boolean canBeDefault() {
    return true;
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/helpers/generators/CommunityModeGenerator.kt
package com.ssblur.scriptor.helpers.generators

import com.google.gson.JsonObject
import java.util.*

class CommunityModeGenerator(obj: JsonObject?) : TokenGenerator() {
    override fun generateToken(key: String, parameters: JsonObject?): String {
        val random = Random((key.hashCode() + 0x055b10b0).toLong())
        val builder = StringBuilder()
        val consonantFirst = random.nextBoolean()
        for (i in 0 until 3 + random.nextInt(4)) {
            if (consonantFirst xor (i % 2 == 1)) builder.append(vowelGroup[random.nextInt(vowelGroup.size)])
            else builder.append(consonantGroup[random.nextInt(consonantGroup.size)])
        }

        return builder.toString()
    }

    override fun canBeDefault(): Boolean {
        return true
    }

    companion object {
        protected var consonantGroup: Array<String> = arrayOf(
            "b",
            "c",
            "d",
            "e",
            "f",
            "g",
            "h",
            "j",
            "k",
            "l",
            "m",
            "n",
            "p",
            "r",
            "s",
            "t",
            "v",
            "w",
            "x",
            "z",
            "st",
            "th",
            "ld",
            "qu",
            "pr",
            "lk",
            "mm",
            "pp",
            "ck",
            "rd",
            "pt"
        )
        protected var vowelGroup: Array<String> = arrayOf(
            "a",
            "e",
            "i",
            "o",
            "u",
            "ou",
            "ao",
            "io",
            "ee"
        )
    }
}
